_posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd

```diff
@@ -0,0 +1,681 @@
+
+As an aside, the model specification here ends up shorter than in _Stan_ as `tfd_binomial` optionally takes logits as parameters.
+
+As with every TFP distribution, a quick functionality check is sampling from the model:
+
+```{r}
+# sample a batch of 2 values 
+# we get samples for every distribution in the model
+s <- m1 %>% tfd_sample(2)
+```
+
+```
+[[1]]
+Tensor("MultivariateNormalDiag/sample/affine_linear_operator/forward/add:0",
+shape=(2, 48), dtype=float32)
+
+[[2]]
+Tensor("IndependentJointDistributionSequential/sample/Beta/sample/Reshape:0",
+shape=(2, 48), dtype=float32)
+```
+
+and computing log probabilities:
+
+```{r}
+# we should get only the overall log probability of the model
+m1 %>% tfd_log_prob(s)
+```
+
+```
+t[[1]]
+Tensor("MultivariateNormalDiag/sample/affine_linear_operator/forward/add:0",
+shape=(2, 48), dtype=float32)
+
+[[2]]
+Tensor("IndependentJointDistributionSequential/sample/Beta/sample/Reshape:0",
+shape=(2, 48), dtype=float32)
+```
+
+You might be wondering why we are not using eager execution, something we normally like to do.
+The reason is that as of today, the sampling code must be executed on the graph in order to be run in parallel -- which definitely is something we'll want to do when with a set of Markov chains. In Python, this can be achieved by annotating the sampling function with `@tf.function`, a construct that as of this writing is not available from R.
+
+Now let's see how we can sample from this model using Hamiltonian Monte Carlo.
+
+## Running Hamiltonian Monte Carlo in TFP
+
+We define a Hamiltonian Monte Carlo kernel with dynamic step size adaptation based on a desired acceptance probability.
+
+```{r}
+# number of steps to run burnin
+n_burnin <- 500
+
+# optimization target is the likelihood of the logits given the data
+logprob <- function(l)
+  m1 %>% tfd_log_prob(list(l, n_surviving))
+
+hmc <- mcmc_hamiltonian_monte_carlo(
+  target_log_prob_fn = logprob,
+  num_leapfrog_steps = 3,
+  step_size = 0.1,
+) %>%
+  mcmc_simple_step_size_adaptation(
+    target_accept_prob = 0.8,
+    num_adaptation_steps = n_burnin
+  )
+```
+
+We then run the sampler, passing in an initial state. If we want to run $n$ chains, that state has to be of length $n$, for every parameter in the model (here we have just one).
+
+The sampling function, `mcmc_sample_chain`, may optionally be passed a `trace_fn` that tells TFP which kinds of meta information to save. Here we save acceptance ratios and step sizes.
```

Comment:
Again, I try to link to function reference on first mention, but definitely your call!

Reply:
totally makes sense, esp. with such a new pkg...